modules/explorer-core/src/main/scala/org/ergoplatform/explorer/db/repositories/DexOrdersRepo.scala

```diff
@@ -0,0 +1,157 @@
+      tokenId: TokenId
+    ): Stream[D, DexSellOrderOutput] =
+      QS.getMainUnspentSellOrderByTokenId(
+          tokenId,
+          sellContractTemplate,
+          0,
+          Int.MaxValue
+        )
+        .stream
+        .map(eOut =>
+          DexSellOrderOutput(
+            eOut,
+            getTokenPriceFromSellOrderTree(eOut.output.ergoTree).get
+          )
+        )
+        .translate(liftK)
+
+    /** Get all unspent main-chain DEX buy orders
+      */
+    override def getAllMainUnspentBuyOrderByTokenId(
+      tokenId: TokenId
+    ): Stream[D, DexBuyOrderOutput] =
+      QS.getMainUnspentBuyOrderByTokenId(tokenId, buyContractTemplate, 0, Int.MaxValue)
+        .stream
+        .map(eOut =>
+          DexBuyOrderOutput(eOut, getTokenInfoFromBuyOrderTree(eOut.output.ergoTree).get)
+        )
+        .translate(liftK)
+
+  }
+
+  /** template for the buyer contract as of http://github.com/ScorexFoundation/sigmastate-interpreter/blob/42e55cbfd093252b8005e4607970764dd6610cbe/contract-verification/src/main/scala/sigmastate/verification/contract/AssetsAtomicExchange.scala#L33-L45
+    */
+  val sellContractTemplate: HexString = HexString
+    .fromString[Try](
+      "eb027300d1eded91b1a57301e6c6b2a5730200040ed801d60193e4c6b2a5730300040ec5a7eded92c1b2a57304007305720193c2b2a5730600d07307"
+    )
+    .get
+
+  /** template for the buyer contract as of http://github.com/ScorexFoundation/sigmastate-interpreter/blob/42e55cbfd093252b8005e4607970764dd6610cbe/contract-verification/src/main/scala/sigmastate/verification/contract/AssetsAtomicExchange.scala#L12-L32
+    */
+  val buyContractTemplate: HexString = HexString
+    .fromString[Try](
+      "eb027300d1eded91b1a57301e6c6b2a5730200040ed803d601e4c6b2a5730300020c4d0ed602eded91b172017304938cb27201730500017306928cb27201730700027308d60393e4c6b2a5730900040ec5a7eded720293c2b2a5730a00d0730b7203"
+    )
+    .get
+
+  private val treeSerializer: ErgoTreeSerializer = new ErgoTreeSerializer
+
+  def getTokenPriceFromSellOrderTree(ergoTree: HexString): Try[Long] =
+    Base16
+      .decode(ergoTree.unwrapped)
+      .flatMap { bytes =>
+        val tree = treeSerializer.deserializeErgoTree(bytes)
+        tree.constants
+        // assuming buyer contract from http://github.com/ScorexFoundation/sigmastate-interpreter/blob/42e55cbfd093252b8005e4607970764dd6610cbe/contract-verification/src/main/scala/sigmastate/verification/contract/AssetsAtomicExchange.scala#L33-L45
+          .lift(5)
+          .collect {
+            case Values.ConstantNode(value, SLong) =>
+              Success(value.asInstanceOf[Long])
+          }
+          .getOrElse(
+            Failure(
+              DexSellOrderAttributesFailed(
+                s"Cannot find tokenPrice in sell order ergo tree $ergoTree"
+              )
+            )
+          )
+      }
+
+  case class TokenInfo(tokenId: TokenId, amount: Long)
+
+  def getTokenInfoFromBuyOrderTree(ergoTree: HexString): Try[TokenInfo] =
```

Comment:
Please move this static helper functions into some utils module

Reply:
Its also better not to commit ourselves to `Try` and abstract it as `F[_]: ContravariantRaise[*[_], DexBuyOrderAttributesFailed]`

Reply:
Thanks! Done. It took me quite some time, but it was fun. :)